Feat/flight sql client

jeremyosterhoudt opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->

Commit:
initial commit

Commit:
feat(sqlclient): implemented methods

**GetExportedKeys**
    - Retrieves a description about the foreign key columns that reference the primary key columns of the given table.
- **GetExportedKeysSchema**
    - Get the exported keys schema from the server.
- **GetImportedKeys**
    - Retrieves the foreign key columns for the given table.
- **GetImportedKeysSchema**
    - Get the imported keys schema from the server.
- **GetCrossReference**
    - Retrieves a description of the foreign key columns in the given foreign key table that reference the primary key or the columns representing a unique constraint of the parent table.
- **GetCrossReferenceSchema**
    - Get the cross reference schema from the server.
- **GetTableTypes**
    - Request a list of table types.
- **GetTableTypesSchema**
    - Get the table types schema from the server.
- **GetXdbcTypeInfo**
    - Request the information about all the data types supported.
- **GetXdbcTypeInfo (with data_type parameter)**
    - Request the information about a specific data type supported.
- **GetXdbcTypeInfoSchema**
    - Get the type info schema from the server.
- **GetSqlInfo**
    - Request a list of SQL information.
- **GetSqlInfoSchema**

HackPoint committed:
Merge branch 'apache:main' into feat/flight-sql-client

HackPoint committed:
Merge branch 'apache:main' into feat/flight-sql-client

Commit:
Merge branch 'refs/heads/main' into feat/flight-sql-client

Commit:
feat: cancel functionality

Commit:
feat: cancel flight info

Commit:
feat: cancel query

HackPoint committed:
Merge branch 'apache:main' into feat/flight-sql-client

Commit:
test(FlightSqlClient): infra and initial testing

HackPoint committed:
Merge branch 'apache:main' into feat/flight-sql-client

Commit:
Merge remote-tracking branch 'origin/feat/flight-sql-client' into feat/flight-sql-client

Commit:
refactor: FlightSqlClient

Commit:
Merge branch 'main' into feat/flight-sql-client

Commit:
feat(test): GetFlightInfo

Commit:
refactor: Flight sql testing

Commit:
testing: flightsql

Commit:
rollback: FlightTests

Commit:
Merge branch 'main' into feat/flight-sql-client

# Conflicts:
#	csharp/src/Apache.Arrow.Flight/Client/FlightClient.cs
#	csharp/test/Apache.Arrow.Flight.Tests/FlightTests.cs

Commit:
test: flight client

Commit:
testing: FlightSqlClient

Commit:
testing: CancelQuery

Commit:
test: DoPut

HackPoint committed:
Merge branch 'apache:main' into feat/flight-sql-client

jeremyosterhoudt commented on cpp/CMakeLists.txt:
```diff
@@ -415,9 +415,9 @@ else()
-if(NOT ARROW_BUILD_EXAMPLES)
-  set(NO_EXAMPLES 1)
-endif()
+#if(NOT ARROW_BUILD_EXAMPLES)
```
I'm guessing this is just playing with C++ and is safe to revert?

HackPoint replied:
This is the games that you and me did when we tried to run c++ server.

I reverted it.

jeremyosterhoudt commented on cpp/src/arrow/flight/sql/CMakeLists.txt:
```diff
@@ -94,7 +94,7 @@ endif()
 
 # Build test server for unit tests
-if(ARROW_BUILD_TESTS OR ARROW_BUILD_EXAMPLES)
+#if(ARROW_BUILD_TESTS OR ARROW_BUILD_EXAMPLES)
```
I'm guessing this is just playing with C++ and is safe to revert?

HackPoint replied:
reverted this change

jeremyosterhoudt commented on cpp/src/arrow/flight/sql/client.h:
```diff
@@ -443,7 +443,7 @@ class ARROW_FLIGHT_SQL_EXPORT PreparedStatement {
                     std::shared_ptr<Schema> dataset_schema,
                     std::shared_ptr<Schema> parameter_schema);
-
+                                                                                                                                                                                                                                             
```
I'm guessing this is just playing with C++ and is safe to revert?

HackPoint replied:
this change was made because of the merge, haven't changed it.

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.IntegrationTest/Apache.Arrow.Flight.Sql.IntegrationTest.csproj:
```diff
@@ -0,0 +1,15 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
I'm not sure if the Arrow teams is going to be OK with a program like this.  I leave it up to you if you want to keep it

HackPoint replied:
removed this project, it was only for testing against our client and test implemented code.

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.IntegrationTest/Program.cs:
```diff
@@ -0,0 +1,425 @@
+/*using Apache.Arrow;
```
I'm not sure if the Arrow teams is going to be OK with a program like this. I leave it up to you if you want to keep it

HackPoint replied:
removed this project, it was only for testing against our client and test implemented code.

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/FlightSqlHolder.cs:
```diff
@@ -0,0 +1,59 @@
+using System.Collections.Generic;
```
Can this holder extend the `FlightHolder` or using composition with the `FlightHolder`?  They look very similar

HackPoint replied:
Didn't wanted to reference the testing project into this project. 

> Can this holder extend the `FlightHolder` or using composition with the `FlightHolder`? They look very similar

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/Apache.Arrow.Flight.Sql.TestWeb.csproj:
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
```
I would suggest seeing if the `Apache.Arrow.Flight.TestWeb` can be used as a dependency and augment what is needed since a lot of the code between the two is identical

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/TestFlightSqlServer.cs:
```diff
@@ -0,0 +1,182 @@
+            case "test":
+                await responseStream.WriteAsync(new FlightResult("test data"));
+                break;
+            case "GetPrimaryKeys":
```
We should use the FlightSQL constants for these items `SqlAction.CreateRequest`.  For the ones that don't exist like `BeginTransaction` we should add them.

HackPoint replied:
done

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/TestFlightSqlServer.cs:
```diff
@@ -0,0 +1,182 @@
+    public override async Task DoGet(FlightTicket ticket, FlightServerRecordBatchStreamWriter responseStream,
+        ServerCallContext context)
+    {
+        // var flightDescriptor = FlightDescriptor.CreatePathDescriptor(ticket.Ticket.ToStringUtf8());
```
remove commented out line

HackPoint replied:
done

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/TestFlightSqlServer.cs:
```diff
@@ -0,0 +1,182 @@
+            _flightStore.Flights.Add(flightDescriptor, flightHolder);
+        }
+
+        while (await requestStream.MoveNext())
```
`await requestStream.MoveNext().ConfigureAwait(false)`

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/TestFlightSqlServer.cs:
```diff
@@ -0,0 +1,182 @@
+            var batches = flightHolder.GetRecordBatches();
+            foreach (var batch in batches)
+            {
+                await responseStream.WriteAsync(batch.RecordBatch, batch.Metadata);
```
Add `ConfigureAwait(false)`

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/TestFlightSqlServer.cs:
```diff
@@ -0,0 +1,182 @@
+
+                var packedResult = Any.Pack(prepareStatementResponse).Serialize().ToByteArray();
+                var flightResult = new FlightResult(packedResult);
+                await responseStream.WriteAsync(flightResult);
```
Add `ConfigureAwait(false)`

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/TestFlightSqlServer.cs:
```diff
@@ -0,0 +1,182 @@
+                var flightDescriptor = FlightDescriptor.CreateCommandDescriptor("test");
+                var flightInfo = new FlightInfo(schema, flightDescriptor, new List<FlightEndpoint>(), 0, 0);
+                var cancelRequest = new CancelFlightInfoRequest(flightInfo);
+                await responseStream.WriteAsync(new FlightResult(Any.Pack(cancelRequest).Serialize().ToByteArray()));
```
Add `ConfigureAwait(false)`

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/TestFlightSqlServer.cs:
```diff
@@ -0,0 +1,182 @@
+        switch (request.Type)
+        {
+            case "test":
+                await responseStream.WriteAsync(new FlightResult("test data"));
```
Add `ConfigureAwait(false)`

jeremyosterhoudt commented on csharp/Apache.Arrow.Flight.Sql.TestWeb/TestFlightSqlServer.cs:
```diff
@@ -0,0 +1,182 @@
+    public override async Task ListActions(IAsyncStreamWriter<FlightActionType> responseStream,
+        ServerCallContext context)
+    {
+        await responseStream.WriteAsync(new FlightActionType("get", "get a flight"));
```
Add `ConfigureAwait(false)` to all `await` calls

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/CancelFlightInfoRequest.cs:
```diff
@@ -0,0 +1,98 @@
+
+namespace Apache.Arrow.Flight.Sql;
+
+public sealed class CancelFlightInfoRequest : IMessage<CancelFlightInfoRequest>
```
We shouldn't shadow the proto buf class like this.  If we need to extend it we should use a partial class to add what is needed

HackPoint replied:
There's no implementation of `CancelFlightInfoRequest` class thus I need to understand what we have to do with this?
@jeremyosterhoudt

HackPoint replied:
done and implemented and tested

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/CancelFlightInfoResult.cs:
```diff
@@ -0,0 +1,96 @@
+    Unrecognized = -1
+}
+
+public sealed class CancelFlightInfoResult : IMessage<CancelFlightInfoResult>
```
We shouldn't shadow the proto buf class like this. If we need to extend it we should use a partial class to add what is needed

HackPoint replied:
There's no implementation of `CancelFlightInfoResult ` class thus I need to understand what we have to do with this?
@jeremyosterhoudt

jeremyosterhoudt replied:
My concern is we are shadowing the `CancelFlightInfoResult` proto by doing this.  We should just change the class to a partial class and remove the extension of `IMessage<CancelFlightInfoResult>`

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            var action = new FlightAction(SqlAction.CreateRequest, prepareStatementRequest.PackAndSerialize());
+            var call = _client.DoAction(action, options.Headers);
+
+            await foreach (var result in call.ResponseStream.ReadAllAsync())
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+
+                byte[] commandSqlCallPackedAndSerialized = commandSqlCall.PackAndSerialize();
+                var descriptor = FlightDescriptor.CreateCommandDescriptor(commandSqlCallPackedAndSerialized);
+                return await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            var call = DoActionAsync(options, action);
+            long affectedRows = 0;
+
+            await foreach (var result in call)
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+                var descriptor = FlightDescriptor.CreateCommandDescriptor(command.PackAndSerialize());
+                var flightInfo = await GetFlightInfoAsync(options, descriptor);
+                var doGetResult = DoGetAsync(options, flightInfo.Endpoints[0].Ticket);
+                await foreach (var recordBatch in doGetResult)
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+
+        var call = _client.DoAction(action, options.Headers);
+
+        await foreach (var result in call.ResponseStream.ReadAllAsync())
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+    /// <returns>An async enumerable of results</returns>
+    public async IAsyncEnumerable<FlightResult> DoActionAsync(FlightAction action)
+    {
+        await foreach (var result in DoActionAsync(new FlightCallOptions(), action))
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            var action = new FlightAction(SqlAction.CreateRequest, prepareStatementRequest.PackAndSerialize());
+            var call = _client.DoAction(action, options.Headers);
+
+            await foreach (var result in call.ResponseStream.ReadAllAsync())
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var command = new CommandGetCatalogs();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(command.PackAndSerialize());
+            var catalogsInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var commandGetCatalogsSchema = new CommandGetCatalogs();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(commandGetCatalogsSchema.PackAndSerialize());
+            var schemaResult = await GetSchemaAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            var command = new CommandGetDbSchemas();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(command.PackAndSerialize());
+
+            var schemaResult = await GetSchemaAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        }
+
+        var call = _client.GetStream(ticket, options.Headers);
+        await foreach (var recordBatch in call.ResponseStream.ReadAllAsync())
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            };
+            byte[] packedRequest = getPrimaryKeysRequest.PackAndSerialize();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(packedRequest);
+            var flightInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            };
+
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(getExportedKeysRequest.PackAndSerialize());
+            var flightInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var commandGetExportedKeysSchema = new CommandGetExportedKeys();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(commandGetExportedKeysSchema.PackAndSerialize());
+            var schemaResult = await GetSchemaAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            };
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(getImportedKeysRequest.PackAndSerialize());
+
+            var flightInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var commandGetImportedKeysSchema = new CommandGetImportedKeys();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(commandGetImportedKeysSchema.PackAndSerialize());
+            var schemaResult = await GetSchemaAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            };
+
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(commandGetCrossReference.PackAndSerialize());
+            var flightInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var command = new CommandGetTableTypes();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(command.PackAndSerialize());
+            var flightInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var command = new CommandGetTableTypes();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(command.PackAndSerialize());
+            var schemaResult = await GetSchemaAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var command = new CommandGetXdbcTypeInfo { DataType = dataType };
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(command.PackAndSerialize());
+            var flightInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var command = new CommandGetXdbcTypeInfo();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(command.PackAndSerialize());
+            var flightInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var command = new CommandGetXdbcTypeInfo();
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(command.PackAndSerialize());
+            var schemaResult = await GetSchemaAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            var command = new CommandGetSqlInfo();
+            command.Info.AddRange(sqlInfo.ConvertAll(item => (uint)item));
+            var descriptor = FlightDescriptor.CreateCommandDescriptor(command.PackAndSerialize());
+            var flightInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        {
+            var action = new FlightAction("CancelFlightInfo", request.PackAndSerialize());
+            var call = _client.DoAction(action, options.Headers);
+            await foreach (var result in call.ResponseStream.ReadAllAsync())
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            var cancelRequest = new CancelFlightInfoRequest(info);
+            var action = new FlightAction("CancelFlightInfo", cancelRequest.ToByteString());
+            var call = _client.DoAction(action, options.Headers);
+            await foreach (var result in call.ResponseStream.ReadAllAsync())
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            var actionBeginTransaction = new ActionBeginTransactionRequest();
+            var action = new FlightAction("BeginTransaction", actionBeginTransaction.PackAndSerialize());
+            var responseStream = _client.DoAction(action, options.Headers);
+            await foreach (var result in responseStream.ResponseStream.ReadAllAsync())
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+            var action = new FlightAction(SqlAction.CreateRequest, preparedStatementRequest.PackAndSerialize());
+            var call = _client.DoAction(action, options.Headers);
+
+            await foreach (var result in call.ResponseStream.ReadAllAsync())
```
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+                };
+                byte[] commandSqlCallPackedAndSerialized = commandSqlCall.PackAndSerialize();
+                var descriptor = FlightDescriptor.CreateCommandDescriptor(commandSqlCallPackedAndSerialized);
+                var flightInfo = await GetFlightInfoAsync(options, descriptor);
```
Add `ConfigureAwait(false)`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/PreparedStatement.cs:
```diff
@@ -0,0 +1,37 @@
+namespace Apache.Arrow.Flight.Sql;
+
+// TODO: Refactor this to match C++ implementation
+public class PreparedStatement
```
This class looks incomplete, if someone were a `PreparedStatement` it doesn't seem like it would function

HackPoint replied:
done

jeremyosterhoudt commented on csharp/test/Apache.Arrow.Flight.Sql.Tests/TestFlightSqlSever.cs:
```diff
@@ -86,4 +86,5 @@ private RecordBatch MockRecordBatch(string name)
         var schema = new Schema(new List<Field> {new(name, StringType.Default, false)}, System.Array.Empty<KeyValuePair<string, string>>());
         return new RecordBatch(schema, new []{ new StringArray.Builder().Append(name).Build() }, 1);
     }
+
```
Remove empty line

HackPoint replied:
done

jeremyosterhoudt commented on csharp/test/Apache.Arrow.Flight.TestWeb/FlightHolder.cs:
```diff
@@ -30,7 +30,7 @@ public class FlightHolder
         //Not thread safe, but only used in tests
         private readonly List<RecordBatchWithMetadata> _recordBatches = new List<RecordBatchWithMetadata>();
-        
+
```
Revert

HackPoint replied:
https://github.com/apache/arrow/blob/1c1f7f359d61b59e446b0a47e0279728551ed89f/csharp/test/Apache.Arrow.Flight.TestWeb/FlightHolder.cs#L32

No changes were done here.

jeremyosterhoudt commented on csharp/test/Apache.Arrow.Flight.TestWeb/TestFlightServer.cs:
```diff
@@ -40,6 +40,15 @@ public override async Task DoAction(FlightAction request, IAsyncStreamWriter<Fli
                 case "test":
                     await responseStream.WriteAsync(new FlightResult("test data"));
                     break;
+                case "BeginTransaction":
```
Use existing constants where available and add new ones that are standard like `Commit`

HackPoint replied:
done

jeremyosterhoudt commented on csharp/test/Apache.Arrow.Flight.Tests/FlightTests.cs:
```diff
@@ -489,7 +489,7 @@ public async Task EnsureCallRaisesRequestCancelled()
             var cts = new CancellationTokenSource();
             cts.CancelAfter(1);
-            
+
```
Revert

HackPoint replied:
No changes were done:
https://github.com/apache/arrow/blob/1c1f7f359d61b59e446b0a47e0279728551ed89f/csharp/test/Apache.Arrow.Flight.Tests/FlightTests.cs#L490

jeremyosterhoudt commented on csharp/test/Apache.Arrow.Flight.Sql.Tests/FlightSqlTestExtensions.cs:
```diff
@@ -13,6 +13,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+using Apache.Arrow.Flight.Sql.Client;
```
Revert

HackPoint replied:
done

jeremyosterhoudt commented on csharp/src/Apache.Arrow/Ipc/ArrowTypeFlatbufferBuilder.cs:
```diff
@@ -304,7 +304,7 @@ public void Visit(DictionaryType type)
                 type.ValueType.Accept(this);
             }
-            
+
```
Revert

HackPoint replied:
No changes were done here:
https://github.com/apache/arrow/blob/1c1f7f359d61b59e446b0a47e0279728551ed89f/csharp/src/Apache.Arrow/Ipc/ArrowTypeFlatbufferBuilder.cs#L305

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Savepoint.cs:
```diff
@@ -0,0 +1,13 @@
+namespace Apache.Arrow.Flight.Sql;
+
+public class Savepoint
```
I don't see this class used anywhere

HackPoint replied:
Deleted

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+    /// <returns>The number of rows affected by the operation.</returns>
+    public async Task<long> ExecuteUpdateAsync(FlightCallOptions options, string query, Transaction? transaction = null)
+    {
+        transaction ??= NoTransaction();
```
Should this be referenced somewhere?

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+    /// <returns>The FlightInfo describing where to access the dataset.</returns>
+    public async Task<FlightInfo> ExecuteAsync(FlightCallOptions options, string query, Transaction? transaction = null)
+    {
+        transaction ??= NoTransaction();
```
Should this be referenced somewhere?

jeremyosterhoudt commented on csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs:
```diff
@@ -0,0 +1,1211 @@
+        _client = client ?? throw new ArgumentNullException(nameof(client));
+    }
+
+    public static Transaction NoTransaction() => new(null);
```
This should be a static readonly variable to avoid unneeded allocations

HackPoint replied:
done

jeremyosterhoudt reviewed (commented):
Please address the issues flagged (I can't request changes for some reason)

kou committed:
MINOR: [Dev][Archery][Integration] Remove debug prints (#44140)

### Rationale for this change

#44099 included debug prints.

### What changes are included in this PR?

Remove debug prints.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.

Authored-by: Sutou Kouhei <kou@clear-code.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

zanmato1984 committed:
GH-44098: [C++] Add home made _mm256_set_m128i for compilers who are missing it (#44116)



### Rationale for this change

AVX2 intrinsic _mm256_set_m128i is missing in GCC versions <8.0 - this is a GCC bug discussed in [1], causing certain CI build failed.

[1] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80582

### What changes are included in this PR?

Check the GCC version and use a home made replacement if necessary.

### Are these changes tested?

Manually tested.

### Are there any user-facing changes?

None.

* GitHub Issue: #44098

Authored-by: Ruoxi Sun <zanmato1984@gmail.com>
Signed-off-by: Antoine Pitrou <antoine@python.org>

amoeba committed:
MINOR: [Docs][Python] Fix return type in docstring for Array.slice (#44134)

### Rationale for this change

Currently the docstring for Array.slice says it returns a RecordBatch. I don't see how this is possible with the existing code. My guess is that this was a copy-and-paste error back when the Array and RecordBatch slice impls were added in https://issues.apache.org/jira/browse/ARROW-547.

### What changes are included in this PR?

Just a docstring change. I copied the language from the take method so things are consistent.

### Are these changes tested?

No.

### Are there any user-facing changes?

Just docs.

Authored-by: Bryce Mecum <petridish@gmail.com>
Signed-off-by: Bryce Mecum <petridish@gmail.com>

raulcd committed:
GH-44149: [Packaging][CI] Remove references to deprecated Ubuntu bionic (#44150)

### Rationale for this change

References to bionic Ubuntu are deprecated.

### What changes are included in this PR?

Remove a couple of references to bionic Ubuntu from `setup-ubuntu.sh`

### Are these changes tested?

Yes, will run CI to validate.

### Are there any user-facing changes?

No
* GitHub Issue: #44149

Authored-by: Raúl Cumplido <raulcumplido@gmail.com>
Signed-off-by: Jacob Wujciak-Jens <jacob@wujciak.de>

khwilson committed:
GH-43809: [Docs] Update extension type examples to not use UUID (#44120)

### Rationale for this change

UUID extension types were made canonical in #41299 and are getting
native support in C++ and Python in #37298. As such, it makes sense to
provide an alternative user-defined extension type as an example that is
unlikely to become a canonical extension type anytime soon.

After discussion in #43809, we determined a `RationalType` would make
sense.

Please note that this is a redo of #43849 as I made a blunder and
accidentally pushed a branch that was in a wonky state.

### What changes are included in this PR?

A change in several doc locations which reference a `UuidType` extension
type have been changed to a `RationalType`.

For consistency, this PR also changes single quotes (`''`) to double
quotes (`""`) throughout the Python examples that it modifies.

Also, seemingly unrelated to this change, some doctests began failing as
numpy changed the `repr` of `float16`'s between 1.x and 2.x. We have
updated the failing doctest so that it supports both styles.

### Are these changes tested?

These are documentation changes and `archery docker run
conda-python-docs` succeeds locally.

### Are there any user-facing changes?

No.

cc @ianmcook @rok 
* GitHub Issue: #43809

---------

Co-authored-by: Ian Cook <ianmcook@gmail.com>

AlenkaF committed:
GH-37756: [Format][Docs] Document IPC Compression (#43950)

### Rationale for this change

There is no information about buffer compression of the record batch IPC
message in the format docs
(https://arrow.apache.org/docs/format/Columnar.html).

### What changes are included in this PR?

New paragraph is added with basic information about buffer compression
in IPC.

### Are these changes tested?

No, it is only documentation update.

### Are there any user-facing changes?

No, only documentation update.
* GitHub Issue: #37756

---------

Co-authored-by: Ian Cook <ianmcook@gmail.com>
Co-authored-by: Sutou Kouhei <kou@cozmixng.org>
Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

raulcd committed:
GH-43875: [Go][CI] Remove Go related lint configurations (#44144)

### Rationale for this change

The Go implementation is moving to apache/arrow-go from go/ in apache/arrow.

### What changes are included in this PR?

Removing the linting configuration

### Are these changes tested?

Yes

### Are there any user-facing changes?

No
* GitHub Issue: #43875

Authored-by: Raúl Cumplido <raulcumplido@gmail.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

kou committed:
GH-43874: [CI][Integration][Go] Use apache/arrow-go (#44142)

### Rationale for this change

The Go implementation is moving to apache/arrow-go from go/ in apache/arrow.

### What changes are included in this PR?

Use apache/arrow-go for integration test.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.
* GitHub Issue: #43874

Authored-by: Sutou Kouhei <kou@clear-code.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

raulcd committed:
GH-43868: [CI][Python] Skip test that requires PARQUET_TEST_DATA env on emscripten (#43906)

### Rationale for this change

The following PR:
- https://github.com/apache/arrow/pull/41580

Made mandatory for a test the requirement to have `PARQUET_TEST_DATA` env defined.

This is currently not available from `python_test_emscripten.sh` as we require to mount the filesystem for both Node and ChromeDriver.

### What changes are included in this PR?

Skip the test that requires `PARQUET_TEST_DATA` for emscripten.

### Are these changes tested?

Via archery

### Are there any user-facing changes?

No
* GitHub Issue: #43905
* GitHub Issue: #43868

Lead-authored-by: Raúl Cumplido <raulcumplido@gmail.com>
Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

kou committed:
GH-44153: [GLib][FlightRPC] Fix closure annotation (#44154)

### Rationale for this change

It should be added to a callback function not data for the callback function.

### What changes are included in this PR?

Move annotation location.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* GitHub Issue: #44153

Authored-by: Sutou Kouhei <kou@clear-code.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

Commit:
Merge branch 'main' into feat/flight-sql-client

Commit:
chore: protobuf version fix

Commit:
chore: remove commented out commands in CMakeLists.txt

Commit:
chore: adding ConfigureAwait(false) to every async method to TestFlightSqlServer.cs
csharp/Apache.Arrow.Flight.Sql.TestWeb/TestFlightSqlServer.cs
            var batches = flightHolder.GetRecordBatches();
            foreach (var batch in batches)
            {
                await responseStream.WriteAsync(batch.RecordBatch, batch.Metadata);
Add ConfigureAwait(false)

Commit:
chore: attending issues in github
